### PR TITLE
Review Guide

### PR DESCRIPTION
**Pull request by terakilobyte**
I propose adding a review guide. It may not be perfect and we can tweak it over time, but I think it's a good place to start. It came up during a retro (and other team's retros, I've heard) that we didn't have very clear review guidelines.

### CODE REVIEW

File: source/reviews/comments.txt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+secondary goal is improving the skills of authors so that they require less
+and less review over time.
+
+Remember that people learn from reinforcement of what they are doing well and
+not just what they could do better. If you see things you like in the PR,
+comment on those too! Examples: author cleaned up a messy paragraph, added
+new examples or tests, or you as the reviewer learned something from the PR.
+Just as with all comments, include why you liked something, further encouraging
+the author to continue good practices.
+
```
Comment (ccho-mongodb) [inline source/reviews/comments.txt:96]: I wonder if this section can -- similar to the "Writing Guidelines" in the style guide -- be direct in what guidance you're providing and then explain why. Otherwise, it can be difficult to locate a specific piece of guidance and what situation it applies to. E.g. "Do X. <Here's why...>"

File: source/reviews/comments.txt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+Accepting Explanations
+++++++++++++++++++++++
+
+If you ask a author to explain content that you don't understand,
+that should usually result in them **rewriting the content more clearly**.
+
+**Explanations written only in the code review tool are not helpful to future readers.**
+They are acceptable only in a few circumstances, such as when
+you are reviewing an area you are not familiar with and the author
+explains something that normal readers of the code would have already known.
```
Comment (ccho-mongodb) [inline source/reviews/comments.txt:106]: Just want to make sure you identify "code" as equivalent to "content" somewhere and otherwise distinguish between source code and non-source code documentation.

**Review by ccho-mongodb — COMMENTED** [review]
Only got the chance to look at comments so far, will continue later.

**Review by ccho-mongodb — COMMENTED** [review]
Still in progress, but thought I should release these pending comments from before Thanksgiving break.

File: source/reviews/standard.txt [hunk 3]
```diff
@@ -0,0 +1,90 @@
+======================
+The Standard of Review
```
Comment (ccho-mongodb) [inline source/reviews/standard.txt:2]: Contents of this section seem more like "Review Process" rather than a review since it involves both the reviewer and reviewee's tasks. Maybe "Standards for the Review Process" could work.

File: source/reviews/standard.txt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+======================
+
+.. _pr-standard:
+
+Purpose
+-------
+
+The primary purpose of a review is to make sure that the overall health of MongoDB's
+corpus is improving over time. All of the tools and processes of review are
```
Comment (biniona-mongodb) [inline source/reviews/standard.txt:11]: Comment: I've never heard "corpus" used to describe our content outside of this document. Would it make sense to choose a different word?

I've heard "corpus" only used in the context of computational linguistic for training models. Maybe "content", "collected writings", "docuntation"? Not sure of a better word.

File: source/reviews/standard.txt [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+designed to this end.
+
+In order to accomplish this, a series of trade-offs have to be balanced.
+
+First, authors must be able to **make progress** on their tasks. If you never
+submit an improvement to the corpus, then the corpus never improves. Also,
+if a reviewer makes it very difficult for *any* change to go in, then authors
```
Comment (ccho-mongodb) [inline source/reviews/standard.txt:18]: I'm not sure whether this means this is difficult due to some inaccuracy or refusal by the reviewee, or lack of solutions presented by the reviewer. I think you've recommended providing solutions as the reviewer, so it would be useful to identify whether this refers to specifically.

File: source/reviews/standard.txt [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+-------
+
+The primary purpose of a review is to make sure that the overall health of MongoDB's
+corpus is improving over time. All of the tools and processes of review are
+designed to this end.
+
+In order to accomplish this, a series of trade-offs have to be balanced.
+
+First, authors must be able to **make progress** on their tasks. If you never
+submit an improvement to the corpus, then the corpus never improves. Also,
```
Comment (biniona-mongodb) [inline source/reviews/standard.txt:17]: Question: Could "make progress" be defined? I'm also not sure what "if you never submit an improvement..." means in this context. Does it relate to "making progress"? Is this saying, "approved PRs are the way the MongoDB collection of writings improves, and thus you should favor approving PRs"?

File: source/reviews/standard.txt [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+if a reviewer makes it very difficult for *any* change to go in, then authors
+are disincentivized to make improvements in the future.
+
+On the other hand, it is the duty of the reviewer to make sure that each
+PR is of such a quality that the overall content health of their corpus
```
Comment (ccho-mongodb) [inline source/reviews/standard.txt:22]: I noticed the term "content health", but haven't seen a definition in this PR. Could this be defined somewhere so it's clear what that means?

File: source/reviews/standard.txt [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+Also, a reviewer has ownership and responsibility over the code they are
+reviewing. They want to ensure that the corpus stays consistent and maintainable.
+
+The Prime Directive
+-------------------
+
+In general, reviewers should favor approving a PR once it is in a state
+where it definitely improves the overall content health of the area
+being worked on, even if the PR isn't perfect.
```
Comment (biniona-mongodb) [inline source/reviews/standard.txt:36]: Comment: This "definitely improves the overall content health..." seems restrictive. I think "most likely improves" is more inline with "favor approving a PR"

File: source/reviews/comments.txt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+==========================================
+
+Summary
+-------
+
+- Be kind.
+- Explain your reasoning.
+- Balance giving explicit directions with just pointing out problems and letting
+  the author decide.
+- Encourage authors to simplify code or add code comments instead
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:12]: Comment: This seems to go against guideline of "don't include code comments" I've received throughout reviews.

```suggestion
- Encourage authors to simplify code instead
```

File: source/reviews/standard.txt [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+ability to deny approval.
+
+A key point here is that there is no such thing as "perfect" content; there is
+only *better* content. Reviewers should not require the author to polish every tiny
+piece of a PR before granting approval. Rather, the reviewer should balance out
+the need to make forward progress compared to the importance of the changes they
+are suggesting. Instead of seeking perfection, what a reviewer should seek is
+*continuous improvement*. A PR that, as a whole, improves the maintainability,
+readability, and understandability of the corpus shouldn't be delayed for days
+or weeks because it isn't "perfect."
```
Comment (ccho-mongodb) [inline source/reviews/standard.txt:52]: I think improving content health and the goal of improving it is described in multiple places in different forms. Maybe consolidating the definition and single mentions per section would be sufficient.

File: source/reviews/comments.txt [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+  of just explaining the complexity to you.
+- Encourage authors to simplify explanations when appropriate. We're all
+  smart, but smart words take more time to read. Readers prefer
+  conciseness and the simplest explanation in most situations.
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:16]: Comment: I think this guideline conflates "concise" with "simple". I believe this ideas should be discussed separately ([short snake code to illustrate point](https://github.com/guckstift/shortest-js-snake/blob/master/index.html))

File: source/reviews/comments.txt [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+Label Your Comments
++++++++++++++++++++
+
+Label comments similar to `Conventional Comments <https://conventionalcomments.org/>`__
+so the author understands this is a suggestion, issue, question, or comment.
+
+- **No "nits"**.
+- Issue: Something is logically, factually, or grammatically incorrect and needs
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:25]: Question: Is a style guide violation an issue?

File: source/reviews/standard.txt [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+---------
+
+Reviews have an important function in teaching authors something new. It's
+always fine to leave comments that help an author learn something new. Sharing
+knowledge is part of improving the content health of a corpus over time.
+
+Principles
+----------
+
+- Technical facts and data overrule opinions and personal preferences.
```
Comment (ccho-mongodb) [inline source/reviews/standard.txt:73]: Should we also add "requirements" from external authors, based on prior PR reviews to this item?

File: source/reviews/comments.txt [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+.. example::
+
+   Bad: \*\*question\*\* Why did **you** use a semicolon to join two unrelated
+   clauses?
+
+   Good: \*\*suggestion: (blocking)\*\* Semicolons should only be used to
+   join two related independent clauses in place of a comma and coordinating
+   conjunction. Here's how you'd use a semicolon in this instance...
+
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:66]: Suggestion: I've found searching for the word "you" to be helpful

```suggestion
.. tip:: Check for "you"

   A strategy to check for commenting on content vs. author is to run "cmnd+f" for the word "you".
```

File: source/reviews/comments.txt [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+One thing you'll notice about the "good" example from above is that it helps the
+author understand *why* you are making your comment. You don't always need to
+include this information in your review comments, but sometimes it's appropriate
+to give a bit more explanation around your intent, the best practice you're
+following, or how your suggestion improves code health.
+
+Giving Guidance
++++++++++++++++
+
+**In general it is the author's responsibility to fix a PR, not the reviewer's.**
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:80]: Suggestion: Should this be made explicit? If not, should guidelines be given when this should/shouldn't be followed?

```suggestion
**It is the author's responsibility to fix a PR, not the reviewer's.**
```

File: source/reviews/speed.txt [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+Why Reviews Should Be Fast
+--------------------------
+
+**At MongoDB, we optimize for the speed at which a team of authors can produce
+content together**, as opposed to optimizing for the speed at which an
+individual author can write content. The speed of an individual is
+important, it's just not *as* important as the velocity of the entire team.
+
+When code reviews are slow, several things happen:
```
Comment (ccho-mongodb) [inline source/reviews/speed.txt:13]: I think it's worth mentioning a number of factors that cause reviews to be slow since the reviewee can often mitigate the risk.
- Size of PR
- Reaching conclusive/acceptable solutions to PR comments quickly
- Adhering to the scope of the ticket
- Describing the changes made and any scope creep in the ticket or PR
- Proofreading and preparing PRs according to PR template guidelines

File: source/reviews/comments.txt [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+should strike an appropriate balance between pointing out problems and providing
+direct guidance. Pointing out problems and letting the developer make a decision
+often helps the author learn, and makes it easier to do reviews. It also
+can result in a better solution, because the author is closer to the content
+than the reviewer is.
+
+That said, sometimes direct instructions, suggestions, or even content are more
+helpful. The primary goal of review is to get the best PR possible. A
+secondary goal is improving the skills of authors so that they require less
+and less review over time.
```
Comment (biniona-mongodb) [inline source/reviews/comments.txt:94]: Suggestion: This value judgment seems central to how we conduct reviews. Should this be stated clearly outside of this section in a prominent location?

File: source/reviews/pushback.txt [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+This isn't because authors are irresponsible, but because they have a lot of work
+to do and the cleanup gets lost or forgotten in the press of other work. Thus, it
+is usually best to insist that the author clean up their PR *now*, before the
+content is in the corpus and "done." Letting people "clean things up later" is a
+common way for corpus to degenerate.
+
+If a PR introduces new sections, it must be cleaned up before submission
+unless it is an emergency. If the PR exposes surrounding problems and they
+can't be addressed right now, the author should file a ticket for the
+cleanup and assign it to themselves so that it doesn't get lost.
```
Comment (zach-carr) [inline source/reviews/pushback.txt:63]: I think that if an author is going to defer some cleanup, they should create & link the new ticket in that PR. That helps with the accountability/ organization and also leaves a helpful trail for anyone who revisits the PR.

File: source/reviews/speed.txt [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+Speed of Reviews
+================
+
+Why Reviews Should Be Fast
+--------------------------
+
+**At MongoDB, we optimize for the speed at which a team of authors can produce
+content together**, as opposed to optimizing for the speed at which an
+individual author can write content. The speed of an individual is
+important, it's just not *as* important as the velocity of the entire team.
```
Comment (ccho-mongodb) [inline source/reviews/speed.txt:11]: I think there's a balance here between providing feedback vs omitting it to optimize for speed. It's still important to provide feedback when something isn't right.

File: source/reviews/pushback.txt [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+about this, and will immediately write a follow-up PR that fixes the issue.
+However, experience shows that as more time passes after a authors writes the
+original PR, the less likely this clean up is to happen. In fact, unless the
+author does the clean up *immediately* after the present PR, it often never happens.
+This isn't because authors are irresponsible, but because they have a lot of work
+to do and the cleanup gets lost or forgotten in the press of other work. Thus, it
+is usually best to insist that the author clean up their PR *now*, before the
+content is in the corpus and "done." Letting people "clean things up later" is a
+common way for corpus to degenerate.
+
```
Comment (biniona-mongodb) [inline source/reviews/pushback.txt:59]: Suggestion: I think an exception should be made for waterfall projects where lots of unresolved issues make "fix later" often the best solution for avoiding duplicate work

File: source/reviews/respect.txt [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+  intimidate the reviewee into doing what you want, it’s not helpful in the long run -
+  they will feel incapable, and there is not much info in there to help them improve.
+  “This is a good start, but it could use some work” or “This needs some cleanup”
+  are nicer ways of saying it. Discuss the content, not the person.
+
+- Don't discourage tool use. If people use the automated formatter, be grateful
+  they‘re willing to give up the power of formatting in favor of ensuring consistency.
+  Think carefully before enforcing your own preferences over it.
+
+- Don't bike shed. Always ask yourself if this decision really matters in the long
```
Comment (zach-carr) [inline source/reviews/respect.txt:82]: I haven't heard "bike shed," it'd be helpful for me if the meaning were written out

Comment (biniona-mongodb) [inline source/reviews/respect.txt:82]: Suggestion: I've never heard "bikeshed". Could this link to a definition
```suggestion
- Don't [bikeshed](https://en.wiktionary.org/wiki/bikeshed). Always ask yourself if this decision really matters in the long
```

File: source/reviews/speed.txt [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+returning from a meeting, coming back from the restroom, etc.
+
+Fast Responses
+--------------
+
+When we talk about the speed of reviews, it is the *response* time that we
+are concerned with, as opposed to how long it takes a PR to get through the
+whole review and be accepted. The whole process should also be fast, ideally,
+but it's even more important for the *individual responses* to come quickly
+than it is for the whole process to happen rapidly.
```
Comment (ccho-mongodb) [inline source/reviews/speed.txt:67]: I think it should be both: as a potential reviewer, both mentioning you'll review it as well as completing the review is important. I think you should only pick it up when you're ready to review and that the review should not take long. Otherwise, it can turn into a long-running transaction during which nobody else that might be free will touch it.

File: source/reviews/comments.txt [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+  smart, but smart words take more time to read. Readers prefer
+  conciseness and the simplest explanation in most situations.
+
+Label Your Comments
++++++++++++++++++++
+
+Label comments similar to `Conventional Comments <https://conventionalcomments.org/>`__
+so the author understands this is a suggestion, issue, question, or comment.
+
+- **No "nits"**.
```
Comment (zach-carr) [inline source/reviews/comments.txt:24]: This surprises me because I label little improvements "nit" all the time - I think of it as "optional, maybe a style issue, but I do think this would be better - change if you agree".

It sounds like these would just be "suggestions," but I think it can be helpful to convey the weight of the suggestion sometimes.

Just thinking out loud, this system is fine with me.

File: source/reviews/speed.txt [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+
+The reviewer should specify which of these options they intend if it is not
+otherwise clear.
+
+LGTM With Comments is especially worth considering when the author and
+reviewer are in different time zones and otherwise the author would be
+waiting for a whole day just to get "LGTM"
+
+Large PRs
+
```
Comment (ccho-mongodb) [inline source/reviews/speed.txt:111]: As the author, your responsibility should be to point the reviewer to and describe all the relevant changes you've made in a complex or large PR.

File: source/reviews/speed.txt [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+    every time the developer makes an update, the complaints tend to disappear.
+    **Most complaints about the code review process are resolved by making the
+    process faster.**
+-   **Content health can be impacted.** When reviews are slow, there is increased
+    pressure to allow developers to submit PRs that are not as good as they
+    could be. Slow reviews also discourage content cleanups, refactorings, and
+    further improvements to existing PRs.
+
+How Fast Should Code Reviews Be?
+--------------------------------
```
Comment (biniona-mongodb) [inline source/reviews/speed.txt:34]: Suggestion: Could a discussion be added about focus time? Personally, I've found that context switching between reviews and ticketed work makes my velocity on both slow.

Is batching reviews at EOD acceptable?

File: source/reviews.txt [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+
+   .. toctree::
+      :titlesonly:
+
+      /reviews/standard
+      /reviews/speed
+      /reviews/comments
+      /reviews/pushback
+      /reviews/respect
+
```
Comment (zach-carr) [inline source/reviews.txt:20]: Not sure where to leave this comment, but do we want to define what requires a review and what doesn't? On the Cloud team, typo fixes and other straightforward changes (at the discretion of the author) don't require review.

File: source/reviews/comments.txt [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+- Issue: Something is logically, factually, or grammatically incorrect and needs
+  to be corrected. Issues are **always** blocking and no PR should be merged with
+  an open issue.
+- Question: You (the reviewer) have a question about something in a PR.
+  **Do not use** questions as in a Socratic style. If there is a logical, factual, or
```
Comment (ccho-mongodb) [inline source/reviews/comments.txt:29]: I'm not sure I know what to add if there's something that doesn't make sense to me, whether it's knowledge of the product or how it's worded. I think questions in Socratic style and genuine questions are difficult to distinguish through a written medium.

File: source/reviews/comments.txt (hunk 11 above)
Comment (terakilobyte) [reply source/reviews/comments.txt:16]: I need to simplify this

File: source/reviews/comments.txt (hunk 12 above)
Comment (terakilobyte) [reply source/reviews/comments.txt:25]: yes

File: source/reviews/comments.txt [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+saying something that might otherwise be upsetting or contentious.
+
+.. example::
+
+   Bad: \*\*question\*\* Why did **you** use a semicolon to join two unrelated
+   clauses?
+
+   Good: \*\*suggestion: (blocking)\*\* Semicolons should only be used to
+   join two related independent clauses in place of a comma and coordinating
+   conjunction. Here's how you'd use a semicolon in this instance...
```
Comment (ccho-mongodb) [inline source/reviews/comments.txt:65]: This contraction is "you should" and therefore contains "you". I think this should be "make your best attempt" rather than "always".

File: source/reviews/pushback.txt [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+this PR in. So, they say they will clean something up in a later PR or capture it
+in a ticket, and thus you should LGTM *this* PR now. Some authors are very good
+about this, and will immediately write a follow-up PR that fixes the issue.
+However, experience shows that as more time passes after a authors writes the
+original PR, the less likely this clean up is to happen. In fact, unless the
+author does the clean up *immediately* after the present PR, it often never happens.
+This isn't because authors are irresponsible, but because they have a lot of work
+to do and the cleanup gets lost or forgotten in the press of other work. Thus, it
+is usually best to insist that the author clean up their PR *now*, before the
+content is in the corpus and "done." Letting people "clean things up later" is a
```
Comment (ccho-mongodb) [inline source/reviews/pushback.txt:57]: I think these guidelines would work better for the author rather than the reviewer. That way, it doesn't introduce potential conflict and doesn't push blame on the reviewer for not being insistent enough when the result is that it falls through the cracks. If the guidelines were clear to the author, it's clear that it was their responsibility to take care of, and not that the reviewer wasn't insistent enough.

File: source/reviews/pushback.txt [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+health, they should continue to advocate for the change if they believe the
+resulting content quality improvement justifies the additional work requested.
+**Improving content health tends to happen in small steps.**
+
+Sometimes it takes a few rounds of explaining a suggestion before it really
+sinks in. Just make sure to always stay polite and let the author know that
+you *hear* what they're saying, you just don't *agree*.
+
+Upsetting Authors
+-----------------
```
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/reviews/pushback.txt:33]: To expand on this comment, I think this section should be replaced with one that explains strategies to handle situations in which the author seems upset by one or more of the reviewer's comments. How should each individual approach these situations? Sometimes communicating outside of the PR threads is a much more effective strategy for building understanding and resolving the issue than assuming they might not be upset or won't be upset for long.

File: source/reviews/author/feedback.txt [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+may know something  that the reviewer
+doesn't, so fill them in and give them more context. Usually you can come to
+some consensus between yourself and the reviewer based on technical facts.
+
+Resolving Conflicts
+-------------------
+
+Your first step in resolving conflicts should always be to try to come to
+consensus with your reviewer. If you can't achieve consensus, see
+[The Standard of Code Review](../reviewer/standard.md), which gives principles
```
Comment (mongodben) [inline source/reviews/author/feedback.txt:77]: need to RST-ify

**Comment by ccho-mongodb** [comment]
Moved to the company wiki.
